airflow/dag_processing/bundles/base.py

```diff
@@ -18,12 +18,224 @@
+        )
+        return sorted(val, key=attrgetter("dt"), reverse=True)[min_versions_to_keep:]
+
+    @staticmethod
+    def _remove_recent(val: list[TrackedBundleVersionInfo]) -> list[TrackedBundleVersionInfo]:
+        age_threshold = conf.getint(
+            section="dag_processor",
+            key="stale_bundle_cleanup_age_threshold",
+            fallback=60 * 60 * 24,
+        )
+        ret = []
+        now = pendulum.now(tz=pendulum.UTC)
+        cutoff = now - timedelta(seconds=age_threshold)
+        for item in val:
+            if item.dt < cutoff:
+                ret.append(item)
+        return ret
+
+    def _find_all_tracking_files(self, bundle_name) -> list[TrackedBundleVersionInfo] | None:
+        tracking_dir = get_bundle_tracking_dir(bundle_name=bundle_name)
+        found: list[TrackedBundleVersionInfo] = []
+        if not tracking_dir.exists():
+            log.debug("bundle usage tracking directory does not exist. tracking_dir=%s", tracking_dir)
+            return None
+        for file in tracking_dir.iterdir():
+            log.debug("found bundle tracking file, path=%s", file)
+            version = file.name
+            dt_str = file.read_text()
+            dt = self._parse_dt(val=dt_str)
+            if not dt:
+                log.error(
+                    "could not parse val as datetime bundle_name=%s val=%s version=%s",
+                    bundle_name,
+                    dt_str,
+                    version,
+                )
+                continue
+            found.append(TrackedBundleVersionInfo(lock_file_path=file, version=version, dt=dt))
+        return found
+
+    @staticmethod
+    def _remove_stale_bundle(bundle_type: str, bundle_name: str, info: TrackedBundleVersionInfo) -> None:
+        bundle_version_path = get_bundle_version_path(
+            bundle_type=bundle_type,
+            bundle_name=bundle_name,
+            version=info.version,
+        )
+
+        def log_info(msg):
+            log.info(
+                "%s bundle_name=%s bundle_version=%s bundle_path=%s lock_file=%s",
+                msg,
+                bundle_name,
+                info.version,
+                bundle_version_path,
+                info.lock_file_path,
+            )
+
+        try:
+            log_info("removing stale bundle.")
+            with open(info.lock_file_path, "a") as f:
+                flock(f, fcntl.LOCK_EX | fcntl.LOCK_NB)  # exclusive lock, do not wait
+                # remove the actual bundle copy
+                shutil.rmtree(bundle_version_path)
+                # remove the lock file
+                os.remove(info.lock_file_path)
+        except BlockingIOError:
+            log_info("could not obtain lock. stale bundle will not be removed.")
+            return
+
+    def _find_candidates(self, found):
+        """Remove the recently used bundles."""
+        candidates = self._remove_last_n(found)
+        candidates = self._remove_recent(candidates)
```

Comment:
Doesn't this mean we don't necessarily keep the last n?

Feels like maybe it should be "keep newer than x or at least the last n"?

Reply:
it's just remove from the candidates list :) yeah i can make that clearer